File: flutter_client/lib/services/webrtc_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import '../models/call.dart';
import '../models/user.dart';
import 'signalr_service.dart';

class WebRTCService extends ChangeNotifier {
  final SignalRService _signalRService;
  
  // WebRTC 状态
  bool _isInitialized = false;
  Call? _currentCall;
  bool _isInCall = false;
  String? _localStreamId;
  String? _remoteStreamId;
  
  // 回调函数
  Function(Call)? onIncomingCall;
  Function(Call)? onCallAccepted;
  Function(Call)? onCallRejected;
  Function(Call)? onCallEnded;
  Function(String)? onConnectionEstablished;
  Function(String)? onConnectionLost;
  Function(String)? onError;

  WebRTCService(this._signalRService) {
    _setupSignalRHandlers();
  }

  // Getters
  bool get isInitialized => _isInitialized;
  Call? get currentCall => _currentCall;
  bool get isInCall => _isInCall;
  String? get localStreamId => _localStreamId;
  String? get remoteStreamId => _remoteStreamId;

  // 初始化WebRTC服务
  Future<void> initialize(String token) async {
    try {
      await _signalRService.connect(token);
      _isInitialized = true;
      notifyListeners();
      print('✅ WebRTC服务初始化成功');
    } catch (e) {
      print('❌ WebRTC服务初始化失败: $e');
      onError?.call('WebRTC服务初始化失败: $e');
    }
  }

  // 设置SignalR事件处理器
  void _setupSignalRHandlers() {
    _signalRService.onIncomingCall = (call) {
      _currentCall = call;
      onIncomingCall?.call(call);
      notifyListeners();
    };

    _signalRService.onCallAccepted = (callId) {
      if (_currentCall != null) {
        _isInCall = true;
        onCallAccepted?.call(_currentCall!);
        notifyListeners();
      }
    };

    _signalRService.onCallRejected = (callId) {
      _currentCall = null;
      _isInCall = false;
      notifyListeners();
    };

    _signalRService.onCallEnded = (callId) {
      _currentCall = null;
      _isInCall = false;
      _localStreamId = null;
      _remoteStreamId = null;
      notifyListeners();
    };
  }

  // 处理WebRTC信令消息
  void _handleWebRTCMessage(Map<String, dynamic> messageData) {
    try {
      final messageType = messageData['type'] as String;
      final callId = messageData['callId'] as String;
      final data = messageData['data'] as String;

      switch (messageType) {
        case 'Offer':
          _handleOffer(callId, data);
          break;
        case 'Answer':
          _handleAnswer(callId, data);
          break;
        case 'IceCandidate':
          _handleIceCandidate(callId, data);
          break;
        default:
          print('⚠️ 未知的WebRTC消息类型: $messageType');
      }
    } catch (e) {
      print('❌ 处理WebRTC消息失败: $e');
      onError?.call('处理WebRTC消息失败: $e');
    }
  }

  // 处理Offer
  void _handleOffer(String callId, String offer) {
    print('📥 收到Offer: $callId');
    // TODO: 实现WebRTC Offer处理
    // 这里需要与具体的WebRTC实现集成
  }

  // 处理Answer
  void _handleAnswer(String callId, String answer) {
    print('📥 收到Answer: $callId');
    // TODO: 实现WebRTC Answer处理
  }

  // 处理ICE候选
  void _handleIceCandidate(String callId, String candidate) {
    print('📥 收到ICE候选: $callId');
    // TODO: 实现ICE候选处理
  }

  // 发起通话
  Future<void> initiateCall(User receiver, CallType callType) async {
    try {
      if (!_isInitialized) {
        throw Exception('WebRTC服务未初始化');
      }

      // 通过SignalR发起通话
      await _signalRService.initiateCall(InitiateCallRequest(
        receiverId: receiver.id,
        callType: callType,
      ));

      print('📞 发起通话: ${receiver.username}');
    } catch (e) {
      print('❌ 发起通话失败: $e');
      onError?.call('发起通话失败: $e');
      rethrow;
    }
  }

  // 应答通话
  Future<void> answerCall(String callId, bool accept) async {
    try {
      if (!_isInitialized) {
        throw Exception('WebRTC服务未初始化');
      }

      // 通过SignalR应答通话
      await _signalRService.answerCall(AnswerCallRequest(
        callId: callId,
        accept: accept,
      ));

      if (accept) {
        _isInCall = true;
      } else {
        _currentCall = null;
        _isInCall = false;
      }
      notifyListeners();

      print('📞 ${accept ? "应答" : "拒绝"}通话: $callId');
    } catch (e) {
      print('❌ 应答通话失败: $e');
      onError?.call('应答通话失败: $e');
      rethrow;
    }
  }

  // 结束通话
  Future<void> endCall() async {
    try {
      if (_currentCall == null) return;

      final callId = _currentCall!.callId;

      // 通过SignalR结束通话
      await _signalRService.endCall(callId);

      _currentCall = null;
      _isInCall = false;
      _localStreamId = null;
      _remoteStreamId = null;
      notifyListeners();

      print('📞 结束通话: $callId');
    } catch (e) {
      print('❌ 结束通话失败: $e');
      onError?.call('结束通话失败: $e');
      rethrow;
    }
  }

  // 断开连接
  Future<void> disconnect() async {
    try {
      await _signalRService.disconnect();
      _isInitialized = false;
      _currentCall = null;
      _isInCall = false;
      _localStreamId = null;
      _remoteStreamId = null;
      notifyListeners();
      print('🔌 WebRTC服务已断开连接');
    } catch (e) {
      print('❌ 断开连接失败: $e');
      onError?.call('断开连接失败: $e');
    }
  }

  @override
  void dispose() {
    disconnect();
    super.dispose();
  }
}

File: website/src/services/webrtc.service.ts
import type { UserSummaryResponse } from './api.service';
import {
  signalRService,
  WebRTCMessageType,
  type WebRTCMessage,
} from './signalr.service';
import { APP_CONFIG } from '../config/app.config';

export const CallType = {
  Voice: 1,
  Video: 2,
} as const;

export type CallType = (typeof CallType)[keyof typeof CallType];
export type CallPeer = UserSummaryResponse & { id: number };

export interface Call {
  callId: string;
  caller: CallPeer;
  receiver: CallPeer;
  callType: CallType;
  status: number;
  startTime: string;
}

class WebRTCService {
  private peerConnection: RTCPeerConnection | null = null;
  private localStream: MediaStream | null = null;
  private remoteStream: MediaStream | null = null;
  private currentCall: Call | null = null;
  private isInCall: boolean = false;
  private pendingIceCandidates: RTCIceCandidateInit[] = [];
  private iceServersPromise: Promise<RTCIceServer[]> | null = null;

  // Callbacks
  onLocalStream?: (stream: MediaStream) => void;
  onRemoteStream?: (stream: MediaStream) => void;
  onCallEnded?: () => void;
  onError?: (error: string) => void;

  get currentCallInfo(): Call | null {
    return this.currentCall;
  }

  get isInCallState(): boolean {
    return this.isInCall;
  }

  // Create peer connection
  private async createPeerConnection(): Promise<RTCPeerConnection> {
    const configuration: RTCConfiguration = {
      iceServers: await this.loadIceServers(),
    };

    const pc = new RTCPeerConnection(configuration);

    // Attach local stream
    if (this.localStream) {
      this.localStream.getTracks().forEach((track) => {
        pc.addTrack(track, this.localStream!);
      });
    }

    // Listen for remote stream
    pc.ontrack = (event) => {
      console.log('Received remote stream');
      const [stream] = event.streams;
      if (stream) {
        this.remoteStream = stream;
      } else {
        this.remoteStream ??= new MediaStream();
        this.remoteStream.addTrack(event.track);
      }
      this.onRemoteStream?.(this.remoteStream);
    };

    // Listen for ICE candidates
    pc.onicecandidate = (event) => {
      if (event.candidate) {
        const candidate = JSON.stringify(event.candidate);
        // Use current callId if available; otherwise temporary value
        const callId = this.currentCall?.callId || '';
        const currentUserId = this.getCurrentUserId();
        
        // Resolve receiver ID
        let receiverId = 0;
        if (this.currentCall) {
          receiverId = this.currentCall.caller.id === currentUserId
            ? this.currentCall.receiver.id
            : this.currentCall.caller.id;
        }
        
        if (callId && receiverId) {
          signalRService.sendWebRTCMessage({
            call_id: callId,
            type: WebRTCMessageType.IceCandidate,
            data: candidate,
            receiver_id: receiverId,
          });
        }
      }
    };

    // Listen for connection state
    pc.onconnectionstatechange = () => {
      console.log('Connection state:', pc.connectionState);
      console.log('ICE state:', pc.iceConnectionState, 'signaling:', pc.signalingState);
      if (pc.connectionState === 'disconnected' || pc.connectionState === 'failed') {
        this.endCall();
      }
    };
    pc.oniceconnectionstatechange = () => {
      console.log('ICE connection state:', pc.iceConnectionState);
    };
    pc.onicecandidateerror = (event) => {
      console.warn('ICE candidate error:', event.errorCode, event.errorText || '');
    };

    return pc;
  }

  private async loadIceServers(): Promise<RTCIceServer[]> {
    if (!this.iceServersPromise) {
      this.iceServersPromise = this.fetchIceServers();
    }
    return this.iceServersPromise;
  }

  private async fetchIceServers(): Promise<RTCIceServer[]> {
    const fallback: RTCIceServer[] = [
      { urls: 'stun:stun.l.google.com:19302' },
      { urls: 'stun:stun1.l.google.com:19302' },
    ];

    try {
      const response = await fetch(`${APP_CONFIG.API_BASE_URL}/api/system/webrtc-config`);
      if (!response.ok) return fallback;

      const payload = await response.json() as { ice_servers?: unknown };
      if (!Array.isArray(payload.ice_servers)) return fallback;

      const iceServers = payload.ice_servers.flatMap((server): RTCIceServer[] => {
        if (!server || typeof server !== 'object') return [];
        const value = server as { urls?: unknown; username?: unknown; credential?: unknown };
        const urls = Array.isArray(value.urls)
          ? value.urls.filter((url): url is string => typeof url === 'string' && url.length > 0)
          : typeof value.urls === 'string' && value.urls.length > 0
            ? value.urls
            : [];
        if (Array.isArray(urls) && urls.length === 0) return [];

        return [{
          urls,
          ...(typeof value.username === 'string' ? { username: value.username } : {}),
          ...(typeof value.credential === 'string' ? { credential: value.credential } : {}),
        }];
      });

      return iceServers.length > 0 ? iceServers : fallback;
    } catch {
      return fallback;
    }
  }

  // Get user media
  async getUserMedia(callType: CallType): Promise<MediaStream> {
    const constraints: MediaStreamConstraints = {
      audio: true,
      video: callType === CallType.Video,
    };

    try {
      const stream = await navigator.mediaDevices.getUserMedia(constraints);
      this.localStream = stream;
      this.onLocalStream?.(stream);
      return stream;
    } catch (error) {
      console.error('Failed to get user media:', error);
      throw error;
    }
  }

  // Initiate call
  async initiateCall(receiverId: number, callType: CallType): Promise<void> {
    try {
      // Get user media
      await this.getUserMedia(callType);

      // Create peer connection
      this.peerConnection = await this.createPeerConnection();

      // Initiate call. Offer is created only after the receiver accepts.
      await signalRService.initiateCall({
        receiver_id: receiverId,
        call_type: callType,
      });
    } catch (error) {
      console.error('Failed to initiate call:', error);
      this.onError?.('Failed to initiate call');
      throw error;
    }
  }

  setCurrentCall(call: Call): void {
    this.currentCall = call;
  }

  async startCallAsCaller(call: Call): Promise<void> {
    this.currentCall = call;
    this.isInCall = true;

    if (!this.localStream) {
      await this.getUserMedia(call.callType);
    }

    if (!this.peerConnection) {
      this.peerConnection = await this.createPeerConnection();
    }

    await signalRService.joinCall(call.callId);

    const offer = await this.peerConnection.createOffer();
    await this.peerConnection.setLocalDescription(offer);

    const receiverId = call.receiver.id;
    await signalRService.sendWebRTCMessage({
      call_id: call.callId,
      type: WebRTCMessageType.Offer,
      data: JSON.stringify(offer),
      receiver_id: receiverId,
    });
  }

  // Accept call
  async acceptCall(call: Call): Promise<void> {
    try {
      this.currentCall = call;
      this.isInCall = true;

      // Get user media
      await this.getUserMedia(call.callType);

      // Create peer connection
        this.peerConnection = await this.createPeerConnection();

      // Join call
      await signalRService.joinCall(call.callId);

      // Answer call
      await signalRService.answerCall(call.callId, true);

      // Do not create answer here
      // Answer should be created in handleWebRTCMessage after receiving Offer
    } catch (error) {
      console.error('Failed to accept call:', error);
      this.onError?.('Failed to accept call');
      throw error;
    }
  }

  // Reject call
  async rejectCall(callId: string): Promise<void> {
    try {
      await signalRService.answerCall(callId, false);
    } catch (error) {
      console.error('Failed to reject call:', error);
    } finally {
      // Ensure resources are cleaned up
      this.cleanup();
    }
  }

  // Handle WebRTC message
  async handleWebRTCMessage(message: WebRTCMessage): Promise<void> {
    try {
      const data = JSON.parse(message.data);

      switch (message.type) {
        case WebRTCMessageType.Offer: {
          // Create peer connection if not created yet (callee side)
          if (!this.peerConnection) {
            // Ensure user media is available
            if (!this.localStream) {
              // Determine call type from currentCall
              if (this.currentCall) {
                await this.getUserMedia(this.currentCall.callType);
              } else {
                // Default to video call if currentCall is missing
                await this.getUserMedia(CallType.Video);
              }
            }
            this.peerConnection = await this.createPeerConnection();
          }

          // Set remote description (offer)
          await this.peerConnection.setRemoteDescription(new RTCSessionDescription(data));
          console.log('Remote offer installed:', {
            signaling: this.peerConnection.signalingState,
            transceivers: this.peerConnection.getTransceivers().map((item) => ({
              direction: item.direction,
              currentDirection: item.currentDirection,
              kind: item.receiver.track.kind,
            })),
          });
          await this.flushPendingIceCandidates();
          
          // Create answer
          const answer = await this.peerConnection.createAnswer();
          await this.peerConnection.setLocalDescription(answer);
          console.log('Local answer created:', {
            signaling: this.peerConnection.signalingState,
            transceivers: this.peerConnection.getTransceivers().map((item) => ({
              direction: item.direction,
              currentDirection: item.currentDirection,
              kind: item.receiver.track.kind,
            })),
          });

          const answerData = JSON.stringify(answer);
          await signalRService.sendWebRTCMessage({
            call_id: message.call_id,
            type: WebRTCMessageType.Answer,
            data: answerData,
            receiver_id: message.sender_id,
          });
          break;
        }

        case WebRTCMessageType.Answer:
          if (!this.peerConnection) {
            console.error('Received answer without peer connection');
            return;
          }
          await this.peerConnection.setRemoteDescription(new RTCSessionDescription(data));
          await this.flushPendingIceCandidates();
          break;

        case WebRTCMessageType.IceCandidate:
          // ICE delivery can race ahead of the offer/answer. WebRTC rejects a
          // candidate until the matching remote description is installed, so
          // preserve it and flush immediately after setRemoteDescription.
          if (!this.peerConnection || !this.peerConnection.remoteDescription) {
            console.warn('Received ICE candidate before remote description; queued for later');
            this.pendingIceCandidates.push(data);
            return;
          }
          await this.peerConnection.addIceCandidate(new RTCIceCandidate(data));
          break;
      }
    } catch (error) {
      console.error('Failed to handle WebRTC message:', error);
      this.onError?.('Failed to handle WebRTC message');
    }
  }

  // End call
  async endCall(): Promise<void> {
    if (this.currentCall) {
      try {
        await signalRService.endCall(this.currentCall.callId);
      } catch (error) {
        console.error('Failed to end call:', error);
      }
    }

    // Ensure resources are cleaned up
    this.cleanup();
    this.onCallEnded?.();
  }

  // Cleanup resources
  private cleanup(): void {
    // Stop local stream
    if (this.localStream) {
      this.localStream.getTracks().forEach((track) => track.stop());
      this.localStream = null;
    }

    // Close peer connection
    if (this.peerConnection) {
      this.peerConnection.close();
      this.peerConnection = null;
    }

    this.currentCall = null;
    this.isInCall = false;
    this.remoteStream = null;
    this.pendingIceCandidates = [];
  }

  private async flushPendingIceCandidates(): Promise<void> {
    if (!this.peerConnection || this.pendingIceCandidates.length === 0) return;

    const candidates = this.pendingIceCandidates;
    this.pendingIceCandidates = [];
    for (const candidate of candidates) {
      await this.peerConnection.addIceCandidate(new RTCIceCandidate(candidate));
    }
  }

  // Get current user ID from localStorage
  private getCurrentUserId(): number {
    const userStr = localStorage.getItem('user');
    if (userStr) {
      const user = JSON.parse(userStr);
      return user.id;
    }
    return 0;
  }
}

export const webRTCService = new WebRTCService();
